benchmark_runner/workloads/bootstorm_vm.py

```diff
@@ -173,99 +173,132 @@ def _verify_single_vm(self, vm_name, retries=5, delay=10):
-                elif self._endpoint_url and not self._google_drive_shared_drive_id:
-                    self.upload_run_artifacts_to_s3()
-                else:
-                    self._save_artifacts_local = True
+    def _verify_vms_in_parallel(self, vm_names):
+        """
+        This method verifies vms in parallel
+        :param vm_names:
+        :return:
+        """
+        failure_vms = []
+        manager = Manager()
+        return_dict = manager.dict()
+
+        # Split vm_names according to self._threads_limit
+        for i in range(0, len(vm_names), self._threads_limit):
+            bulk = vm_names[i:i + self._threads_limit]
+            processes = []
+
+            for vm_name in bulk:
+                p = Process(target=self._verify_single_vm_wrapper, args=(vm_name, return_dict))
+                p.start()
+                processes.append(p)
+
+            for p in processes:
+                p.join()
+
+        # After all processes are done, collect failures
+        for vm_name, status in return_dict.items():
+            if str(status).lower() != 'true':
+                failure_vms.append(vm_name)
+        return failure_vms
+
+    def _verify_virtctl_vms(self):
+            """
+            This method verifies the virtctl SSH login for each VM, either during the upgrade or once for each VM.
+            It prepares the data for ElasticSearch, generates a must-gather in case of an error, and uploads it to Google Drive.
+            """
+            try:
+                vm_names = self._oc._get_all_vm_names()
+                if not vm_names:
+                    raise MissingVMs("No VM names were retrieved from the cluster.")
 
-                # Error log with details of failed VM, for catching all vm errors
-                logger.error(
-                    f"Failed to verify virtctl SSH login for the following VMs: {', '.join(failure_vms)}")
+                upgrade_done = True
+                failure_vms = []  # List to store failed VM names
 
-        except Exception as err:
-            # Save run artifacts logs
-            self.save_error_logs()
-            raise err
+                if self._wait_for_upgrade_version:
+                    logger.info(f"wait for ocp upgrade version: {self._wait_for_upgrade_version}")
+                    upgrade_done = self._oc.get_cluster_status() == f'Cluster version is {self._wait_for_upgrade_version}'
+                    current_wait_time = 0
+
+                    while (self._timeout <= 0 or current_wait_time <= self._timeout) and not upgrade_done:
+                        failure_vms = self._verify_vms_in_parallel(vm_names)
+                        upgrade_done = self._oc.get_cluster_status() == f'Cluster version is {self._wait_for_upgrade_version}'
+
+                        # Sleep 3 sec between each cycle
+                        time.sleep(self.SLEEP_TIME)
+                        current_wait_time += 1  # Increment the wait time
```

Comment:
I'm still concerned about the timeout/server load issue I raised the other day.  It appears to me that you're verifying all of the VMs in parallel and then sleeping for a second.  There are two problems here:

1. You're almost certainly going to find that the actual timeout will be a lot longer than the desired timeout, because the verification step will take much longer than a second whether you're parallelizing them or not.  The right way to do this is to get the timestamp prior to entering the loop and then checking the current time against the timestamp rather than simply incrementing the wait time.
2. Sleeping only 1 second between each mass check will result in a lot of API traffic.

Reply:
Still creates a fair bit of traffic, and the timeout is going to be wrong.  I suggest checking every 5 seconds unless you have a good reason otherwise, and comparing time.time() against the start rather than counting iterations.

Reply:
Agree, I have updated it to 10 sec.